receiver/hostmetricsreceiver/internal/scraper.go

```diff
@@ -39,15 +50,32 @@ type Scraper interface {
 
 // Factory can create a Scraper.
 type Factory interface {
-	// CreateDefaultConfig creates the default configuration for the Scraper.
-	CreateDefaultConfig() Config
+	BaseFactory
 
 	// CreateMetricsScraper creates a scraper based on this config.
 	// If the config is not valid, error will be returned instead.
-	CreateMetricsScraper(
-		ctx context.Context,
-		logger *zap.Logger,
-		cfg Config) (Scraper, error)
+	CreateMetricsScraper(ctx context.Context, logger *zap.Logger, cfg Config) (Scraper, error)
+}
+
+// ResourceScraper gathers metrics from a low-level resource such as
+// a process.
+type ResourceScraper interface {
+	BaseScraper
+
+	// ScrapeMetrics returns relevant scraped metrics per resource.
```

Comment:
No suggestion, just trying to understand the problem better, since this seems to be the key change here.

AFAICT the main difference between `Scraper` and `ResourceScraper` is that Scraper returns metrics for a single fixed resource (host), while `ResourceScraper` dynamically gathers a list of multiple processes, generates a resource per process and returns metrics for each of them.

So in that sense, `Scraper` is just a special case of `ResourceScraper` with exact one (implicit) resource.

Follow-up question: why don't you need to create a separate resource per CPU in `cpuscraper` or per disk in `diskscraper`? What makes the processes unique?

Reply:
1. For processes, the cardinality can be very high (1000s) which may cause issues when storing the metric data. If grouped by resource, the data is sharded against diff root resources (iiuc).
2. We may want to create lots of metrics against the same process (at the moment its just three, but this will presumably be extended), and its easier to group metrics against the same resource than looking up similar label values.

While they wouldn't be impacted by those issues to the same degree, having a resource per cpu, disk, or network interface could make sense.

@bogdandrutu can probably explain the reasoning in more detail.

Reply:
Thanks, that makes perfect sense.